replicate/common/generic-replicator.go

```diff
@@ -0,0 +1,422 @@
+			}
+			key := MustGetKey(objectMeta)
+			logger.WithField("source", key).Infof("Replicated %s to: %v", key, replicatedList)
+		}
+	}
+}
+
+// ResourceAdded checks resources with ReplicateTo or ReplicateFromAnnotation annotation
+func (r *GenericReplicator) ResourceAdded(obj interface{}) {
+	objectMeta := MustGetObjectMeta(obj)
+	cacheKey := MustGetKey(objectMeta)
+	logger := log.WithField("kind", r.Kind).WithField("resource", cacheKey)
+
+	replicas, ok := r.DependencyMap[cacheKey]
+	if ok {
+		logger.Debugf("objectMeta %s has %d dependents", cacheKey, len(replicas))
+		if err := r.updateDependents(obj, replicas); err != nil {
+			logger.WithError(err).
+				Errorf("Failed to update cache for %s: %v", MustGetKey(objectMeta), err)
+		}
+	}
+
+	// Match resources with "replicate-from" annotation
+	source, replicateFrom := objectMeta.Annotations[ReplicateFromAnnotation]
+	if replicateFrom {
+		if err := r.resourceAddedReplicateFrom(source, obj); err != nil {
+			logger.WithError(err).Errorf(
+				"Could not copy %s -> %s: %v",
+				source, MustGetKey(objectMeta), err,
+			)
+		}
+		return
+	}
+
+	// Match resources with "replicate-to" annotation
+	namespaceList, replicateTo := objectMeta.Annotations[ReplicateTo]
+	if replicateTo {
+		if err := r.resourceAddedReplicateTo(obj, namespaceList, nil); err != nil {
+			logger.
+				WithError(err).
+				Errorf(
+					"Could not replicate %s to other namespaces: %+v",
+					MustGetKey(objectMeta), err,
+				)
+		}
+		return
+	}
+}
+
+// resourceAddedReplicateFrom replicates resources with ReplicateFromAnnotation
+func (r *GenericReplicator) resourceAddedReplicateFrom(sourceLocation string, target interface{}) error {
+	cacheKey := MustGetKey(target)
+
+	logger := log.WithField("kind", r.Kind).WithField("source", sourceLocation).WithField("target", cacheKey)
+	logger.Debugf("%s %s is replicated from %s", r.Kind, cacheKey, sourceLocation)
+	v := strings.SplitN(sourceLocation, "/", 2)
+
+	if len(v) < 2 {
+		return errors.Errorf("Invalid source location expected '<namespace>/<name>', got '%s'", sourceLocation)
+	}
+
+	if _, ok := r.DependencyMap[sourceLocation]; !ok {
+		r.DependencyMap[sourceLocation] = make(map[string]interface{})
+	}
+
+	r.DependencyMap[sourceLocation][cacheKey] = nil
+
+	sourceObject, exists, err := r.Store.GetByKey(sourceLocation)
+	if err != nil {
+		return errors.Wrapf(err, "Could not get source %s: %v", sourceLocation, err)
+	} else if !exists {
+		return errors.Errorf("Could not get source %s: does not exist", sourceLocation)
+	}
+
+	if err := r.UpdateFuncs.ReplicateDataFrom(sourceObject, target); err != nil {
+		return errors.Wrapf(err, "Failed to replicate %s target %s -> %s: %v",
+			r.Kind, MustGetKey(sourceObject), cacheKey, err,
+		)
+	}
+
+	return nil
+}
+
+// resourceAddedReplicateFrom replicates resources with ReplicateTo
+func (r *GenericReplicator) resourceAddedReplicateTo(obj interface{}, namespaceList string, targetNamespace *v1.Namespace) error {
```

Comment:
I don't this this function's interface is ideal, since it appears to be designed to be called in very different ways (by supplying 1) a string with a comma-separated list of regexes OR 2) a `*corev1.Namespace` object. Maybe split this into two different methods?

1. `replicateResourceToMatchingNamespaces(obj interface{}, patterns []string) error`, which then in turn calls...
2. `replicateResourceToNamespaces(obj interface{}, targets []*corev1.Namespace) error`

Reply:
I don't see an issue with splitting this.